src/distributions/range.rs

```diff
@@ -193,7 +192,13 @@ macro_rules! range_int_impl {
                     loop {
                         let v: $u_large = Rand::rand(rng, Uniform);
                         if v <= zone {
-                            return self.low.wrapping_add((v % range) as $ty);
+                            // Using a modulus could work just as well as a
+                            // division. With division the result depends more
+                            // on the most significant bits, with modulus on the
+                            // least significant bits. Because many RNG's of
+                            // lower quality are weak in their least significant
+                            // bits, we use division.
+                            return self.low.wrapping_add((v / range) as $ty);
```

Comment:
Maybe you meant `let n = (zone+1) / range; return sample / zone` (avoiding the overflow)?

That means an extra division somewhere in `new_*`. Since `n = (unsigned_max - range + 1) / range + 1` you may be able to compute both remainder and quotient in a single op with something like [`div`](http://en.cppreference.com/w/cpp/numeric/math/div) (I thought there was a Rust equivalent but can't find it). Even then there's an extra copy and addition IIUC so I expect performance will suffer.

So I'm not convinced this is worth it...

Reply:
Yes, I think I was sleeping while doing the math 🙁.

To avoid two divisions I was thinking along these lines in `new_*`:
```rust
let range = (high as $u_large)
            .wrapping_sub(low as $u_large)
            .wrapping_add(1);
let divisor = unsigned_max / range;
let zone = range * divisor;
```

Two subtractions are traded for one multiply, and one modulus for a division. But have to think about it more careful yet.

Reply:
Asleep, I see :D

You need `(unsigned_max + 1) / range` which of course overflows, so you need this:
```rust
let divisor = (unsigned_max - range + 1) / range;
let zone = range * divisor;
let divisor = divisor + 1;
```
Simultaneous division & remainder would probably still be faster (but more CPU dependent maybe).

Reply:
There is a Rust equivalent, on [`num::Integer`](http://rust-num.github.io/num/num/trait.Integer.html#tymethod.div_rem) and as an [unstable intrinsic](https://michael-f-bryan.github.io/rustc-internal-docs/compiler_builtins/int/udiv/fn.__udivmoddi4.html).

Reply:
I have used them together before, LLVM is able to optimize the two operations into one when they are close together.

Reply:
https://github.com/rust-num/num/blob/master/integer/src/lib.rs#L309